src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs

```diff
@@ -0,0 +1,212 @@
+                    BoundStatement payloadIf = factory.If(payloadNullTest, createPayloadCall);
+
+                    ImmutableArray<BoundStatement> newStatements = newMethodBody.Statements.Insert(0, payloadIf);
+                    newMethodBody = newMethodBody.Update(newMethodBody.Locals, newMethodBody.LocalFunctions, newStatements);
+
+                    if (IsTestMethod(method))
+                    {
+                        // If the method is a test method, wrap the body in:
+                        //
+                        // Instrumentation.FlushPayload();
+                        // try
+                        // {
+                        //     ... body ...
+                        // }
+                        // finally
+                        // {
+                        //     Instrumentation.FlushPayload();
+                        // }
+
+                        BoundStatement firstFlush = factory.ExpressionStatement(factory.Call(null, flushPayload));
+                        BoundStatement secondFlush = factory.ExpressionStatement(factory.Call(null, flushPayload));
+                        BoundStatement tryFinally = factory.Try(newMethodBody, ImmutableArray<BoundCatchBlock>.Empty, factory.Block(ImmutableArray.Create(secondFlush)));
+                        newMethodBody = factory.Block(ImmutableArray.Create(firstFlush, tryFinally));
+                    }
+
+                    return newMethodBody;
+                }
+            }
+
+            dynamicAnalysisSpans = ImmutableArray<SourceSpan>.Empty;
+            return methodBody;
+        }
+
+        private static FieldSymbol GetPayloadField(MethodSymbol method, int methodOrdinal, TypeSymbol payloadType, SyntheticBoundNodeFactory factory, CSharpCompilation compilation)
+        {
+            // ToDo PROTOTYPE (https://github.com/dotnet/roslyn/issues/9810):
+            // If the type containing the method is generic, synthesize a helper type and put the payload field there.
+            // If the payload field is part of a generic type, there will be a new instance of the field per instantiation of the generic,
+            // and so the payload field must be a member of another type.
+            NamedTypeSymbol containingType = method.ContainingType;
+
+            SynthesizedFieldSymbol payloadField = new SynthesizedFieldSymbol(containingType, payloadType, GeneratedNames.MakeSynthesizedInstrumentationPayloadFieldName(method, methodOrdinal), isStatic: true);
+            factory.AddField(containingType, payloadField);
+            return payloadField;
+        }
+        
+        private static MethodSymbol GetInstrumentationHelper(CSharpCompilation compilation, WellKnownMember member, DiagnosticBag diagnostics)
+        {
+            MethodSymbol helper = compilation.GetWellKnownTypeMember(member) as MethodSymbol;
+
+            if ((object)helper == null)
```

Comment:
I believe this check is insufficient, and this method duplicates code that already exists. Consider using static helper from the Binder class.